### ISSUE TITLE
ARIA attributes

### ISSUE THREAD

**Issue by wojciechczerniak** [handsontable/handsontable#8967]
### Description

Having grid layout `<table>`-based gives Handsontable advantage and some of accessibility roles are already defined, but given the complexity of grid components there's still a lot that have to be done. This issue gathers most of the gaps that should be filled.

Following proper focus management handsontable/dev-handsontable#383, we have to ensure each focusable element that may be activated or have other state (expanded/collapsed, sorted asc/desc) have correct assistance for screen readers. 

### Grid

- [ ] Mark `role="grid"` with `aria-rowcount`, as we have virtual viewport and it won't be possible to calculate total number of rows otherwise. If total number of rows is unknown (infinity scroll implementation), set `-1` as value to indicate more rows are available and not loaded yet.
- [ ] Mark each `<tr>` with `aria-rowindex` attribute to indicate current scroll position. With `rowspan` set it to first merged row index. Include header rows as well.
- [ ] Mark `role="grid"` with `aria-colcount`, as we have virtual viewport and it won't be possible to calculate total number of columns otherwise. If total number of columns is unknown (infinity scroll implementation), set `-1` as value to indicate more columns are available and not loaded yet.
- [ ] Mark each `<td>` element with `aria-colindex` attribute to indicate current scroll position. With `colspan` set it to first merged column index. Include header columns as well.
- [ ] Since our grid is created with `<table>` element, native `rowspan` and `colspan` attributes are used. No need for `aria-colspan` and `aria-rowspan` , excessive markup.
- [ ] New setting option should be created with grid label. If label is a short string description, it may be inlined with `aria-label` attribute. But if there is a need for a more complicated description, `aria-describedby` should be used and `#elementId` provided instead. Label is required.

### Headers

- [ ] TH within TR element will have `scope="row"` attribute set
- [ ] TH within THEAD will have `scope="col"` attribute set

### Corner

- [ ] If corner has an associated action it will need an inner button that will be activated by the keyboard navigation (see keyboard navigation issue)
- [ ] That button has to have a description for screen readers that says, when activated will select all cells.
- [ ] It can be labeled with button description or `aria-describedby` attribute. Visually hidden to maintain our current visual design

### Nested rows & Nested headers

- [ ] Hierarchical data applies to nested rows only. Main grid element should have `role="treegrid"` instead of `role="grid"`
- [ ] Additional `aria-level="1"` attribute should be applied to nested rows levels
- [ ] Group is further described with `aria-setsize="number"` with number of child `<tr>` items
- [ ] Item `<tr>` in group should be described with `aria-posinset="number"` attribute
- [ ] Nested headers need `<colgroup>` correctly set and, correct `colspan` set, and `scope="colgroup"` set for higher lvl TH elements
- [ ] Nested row headers will need `rowspan` attribute and, `scope="rowgroup"` set for higher lvl TH elements
- [ ] Expand button for rows and columns should be a `<button>` focusable element
- [ ] Expand button needs `aria-expanded="false"` and change this status. Status is set with `aria-expanded="true"` attribute. When collapsible columns or nested rows are enabled, this attribute has to exist and have `true` or `false` value. Missing tag or `undefined` indicates that the area is neither collapsible, nor expandable.
- [ ] Collapse/expand button should have `aria-controls` attribute that references elements controlled by this active element with space-separated `ids` of rows that are affected by this button. Please keep in mind that `ids` should be unique (to handle multiple instances of Handsontable)
- [ ] Expand button needs `aria-label` explaining what will happen when activated

### Sorting

- [ ] For column header that has enabled sorting and it's active in ascending order, we have to add `aria-sort="ascending"` attribute
- [ ] For column header that has enabled sorting and it's active in descending order, we have to add `aria-sort="descending"` attribute
- [ ] For column header that has enabled sorting and it's not active, we should remove the attribute. Default is `aria-sort="none"` , and not needed.
- [ ] ~~For any other sort order, we should have `aria-sort="other"` option available. It does not occur for our default sort states, but with custom sort comparator developer should indicate the kind of returned order~~ Currently support just for 3 options inside plugin and custom sort method.
- [ ] Multi-columns sorting is tricky. ARIA guidelines does not recommend adding `aria-sort` attribute to more than a single header. We will need an additional indicator for priority. Both visually (ie. `<span aria-hidden="true"> 1 </span>` and for screen readers header should have additional `aria-describedby` attribute with generated description that includes priority. Again, as always with `describedby` attribute, it have to be translatable. ie `<th describedby="SortPriorityText"><span id="SortPriorityText">priority 2</span></th>`

// TODO @wszymanski Please take a look at post: https://adrianroselli.com/2021/06/multi-column-sortable-table-experiment.html

### Selection

- [ ] Selected cells should have `aria-selected="true"` attribute set.
- [ ] For column selection, when a column is selected all cells in the column have `aria-selected="true"` attribute. Column headers should also have the attribute.
- [ ] When a row is selected only TR may have `aria-selected="true"` set, but for consistency with columns and ranges we should set this attribute on cells only.
- [ ] When `selectionMode` is set to any option that allows multiple selection, an `aria-multiselectable="true"` should be added to `role="grid"` element. In such case, all `role="gridcell"` elements should have `aria-selected` attribute set, with either `true` or `false` value. No implicit `undefined` should be used is this use case.

### Editing

- [ ] Our grid provides editing functionality, so when this functionality is disabled we should mark it with `aria-readonly="true"` on cell where editing is disabled. By either setting readOnly option, or disabling the default editor.
- [ ] If readOnly is set for the whole component, we should set `aria-readonly="true"` on `aria-role="grid"` role level and skip `role="gridcell"` level `aria-readonly` attributes.
- [ ] Attribute `aria-disabled` should be used with active elements that currently can't be used.
- [ ] When cell is required, it should be marked with `aria-required="true"` attribute

### Validation

- [ ] Invalid cells should have `aria-invalid="true"` attribute set
- [ ] An attribute `aria-errormessage` should be added next to `aria-invalid="true"` with a reference to element that contains error message description. It should contain possible corrections.

### Various

- [ ] Keyboard: Each element that may be activated or focused by keyboard shortcut  can have an additional `aria-keyshortcuts` attribute with shortcuts that are associated with it. Each shortcut is space-delimited and each shortcut keys are delimited by plus sign ("+").

// TODO: @wszymanski We should check whether there won't be too many information about header with the attribute.

- [ ] I18N: `aria-label` and `aria-describedby` will need different content for different translations
- [ ] I18N: `aria-errormessage` will need different content for different translations
- [ ] To correctly read the ARIA attributes we have to maintain the focus on cells, not `<textarea>`. If we have to keep focus on `<textarea>` we have to experiment with dynamic attributes on a single HTML element.

### References

- [https://sarahmhigley.com/writing/grids-part1/](https://sarahmhigley.com/writing/grids-part1/)
- [https://sarahmhigley.com/writing/grids-part2/](https://sarahmhigley.com/writing/grids-part2/)
- [https://adrianroselli.com/2019/09/table-with-expando-rows.html](https://adrianroselli.com/2019/09/table-with-expando-rows.html)
- [https://adrianroselli.com/2021/04/sortable-table-columns.html](https://adrianroselli.com/2021/04/sortable-table-columns.html)
- [https://adrianroselli.com/2021/06/multi-column-sortable-table-experiment.html](https://adrianroselli.com/2021/06/multi-column-sortable-table-experiment.html)
- [https://w3c.github.io/aria-practices/examples/grid/dataGrids.html](https://w3c.github.io/aria-practices/examples/grid/dataGrids.html)
- [https://www.w3.org/WAI/tutorials/tables/](https://www.w3.org/WAI/tutorials/tables/)
- [https://www.w3.org/TR/wai-aria-practices-1.1/#grid](https://www.w3.org/TR/wai-aria-practices-1.1/#grid)

**Comment by AMBudnik** [comment]
Note to myself inform [forum](https://forum.handsontable.com/t/gh-8967-does-handsontable-support-screen-reader-testing/5442)

**Comment by apwete** [comment]
Quick note, maybe I missed it, but the filter buttons do not have a programmatic name.

**Comment by AMBudnik** [comment]
Hi @apwete 

I have good news. We added support for screen readers https://handsontable.com/docs/javascript-data-grid/accessibility/#support-for-screen-readers

Here https://handsontable.com/docs/javascript-data-grid/release-notes/#_14-0-0 you can read more about the changes.
